amlb/datasets/fileutils.py

```diff
@@ -21,21 +32,45 @@ def is_valid_url(url):
 def url_exists(url):
     if not is_valid_url(url):
         return False
-    head_req = Request(url, method='HEAD')
-    try:
-        with urlopen(head_req) as test:
-            return test.status == 200
-    except URLError as e:
-        log.error(f"Cannot access url %s: %s", url, e)
-        return False
+    if not is_s3_url(url):
+        head_req = Request(url, method='HEAD')
+        try:
+            with urlopen(head_req) as test:
+                return test.status == 200
+        except URLError as e:
+            log.error(f"Cannot access url %s: %s", url, e)
+            return False
+    else:
+        import boto3
+        from botocore.errorfactory import ClientError
```

Comment:
feel free to import boto3 in the top of the file, this is an optional requirement in `amlb`.

Reply:
Wouldn't that cause a crash if it is at top of file and is optional if the user hasn't installed boto3? If it is an optional requirement then we wouldn't want to import it until we are sure we need to use it.

Reply:
I don't know why I wrote `optional`, I meant the exact opposite… probalby skipped the `not`.